### ISSUE TITLE
Gradient Boosting Algorithm in R

### ISSUE THREAD

**Issue by juliettetroadec** [juliettetroadec/hello-word#3]
The accuracy of a predictive model can be boosted in two ways: 
- feature engineering 
- boosting algorithms 
People prefer to work with boosting algorithms as it takes less time and produces similar results. There are multiple boosting algorithms like Gradient Boosting, XGBoost, AdaBoost, Gentle Boost etc. Every algorithm has its own underlying mathematics and a slight variation is observed while applying them.

**Bagging** is an approach where you take random samples of data, build learning algorithms and take simple means to find bagging probabilities.
**Boosting** is similar, however the selection of sample is made more intelligently. We subsequently give more and more weight to hard to classify observations.

**1. Easy example**
We want to improve previous model M, from which we observe an accuracy of 80%.
To go further, the solution is to build a new model with new input variables and try better ensemble learners. But there is a much simpler way:
```
Y = M(X) + error
```
Maybe the error isn't a White Noise and has some correlation with Y (outcome value). We may want to develop a model on the error term.
```
error = G(x) + error2
```
Error rate will improve to a higher number, say 84%. Let’s take another step and regress against error2.
```
error2 = H(x) + error3
```
Combining all these equations together:
```
Y = M(x) + G(x) + H(x) + error3
```
The accuracy will probably be higher than 84%. We now want to find accurate weights for each of these equation parts (learners):
```
Y = alpha * M(x) + beta * G(x) + gamma * H(x) + error4
```
The more accurate weights are, the more efficient is the model. This is the main principle of Boosting.

**Issues:**
- Do we really see non white noise error in regression/classification equations? If not, how can we even use this algorithm?
- if this is possible, why not get near 100% accuracy?

Visualization:
![capture d ecran 2016-11-13 a 17 35 11](https://cloud.githubusercontent.com/assets/17045683/20247243/955dadb0-a9c7-11e6-83d5-1c4d8a0d3940.png)
Starting with the right box: we see one vertical line which becomes the first week learner. Now in total we have 3/10 mis-classified observations. We now start giving higher weights to 3 plus mis-classified observations. Now, it becomes very important to classify them right. Hence, the vertical line towards right edge. We repeat this process and then combine each of the learner in appropriate weights.

**2. Practice**
We first combine train and test set to transform variables
```
library(caret)
rm(list=ls())
library(Metrics)
complete <- read.csv("complete_data.csv", stringsAsFactors = TRUE)
train <- complete[complete$Train == 1,]
score <- complete[complete$Train != 1,] # contribution dataset

set.seed(999)
ind <- sample(2, nrow(train), replace=T, prob=c(0.60,0.40))
trainData<-train[ind==1,] 
# we split the train dataset into a train set and test set to tun the model
testData <- train[ind==2,]

set.seed(999)
ind1 <- sample(2, nrow(testData), replace=T, prob=c(0.50,0.50))
trainData_ens1<-testData[ind1==1,]
testData_ens1 <- testData[ind1==2,]

table(testData_ens1$Disbursed)[2]/nrow(testData_ens1)
# Response Rate of 9.052%
```

We then build a GBM model

```
fitControl <- trainControl(method = "repeatedcv", number = 4, repeats = 4) # error control
trainData$outcome1 <- ifelse(trainData$Disbursed == 1, "Yes","No") # outcome variable

set.seed(33)
gbmFit1 <- train(as.factor(outcome1) ~ ., data = trainData[,-26], method = "gbm", 
                            trControl = fitControl, verbose = FALSE)
```

And then study predictions:

```
# Probabilities
gbm_dev <- predict(gbmFit1, trainData, type= "prob")[,2] 
gbm_ITV1 <- predict(gbmFit1, trainData_ens1,type= "prob")[,2] 
gbm_ITV2 <- predict(gbmFit1, testData_ens1,type= "prob")[,2]
# AUC score
auc(trainData$Disbursed,gbm_dev)
auc(trainData_ens1$Disbursed,gbm_ITV1)
auc(testData_ens1$Disbursed,gbm_ITV2)
```

GBM is the most widely used algorithm. XGBoost is another faster version of boosting learner (other tutorial on XGBoost).

**3. Introduction to Gradient Boosting & GBM package in R**

**What is GBM**:
- predictive modeling algorithm (classification and regression)
- decision trees as a basis
- gradient boosted: iteratively solves residuals
- stochastic
![capture d ecran 2016-11-13 a 17 54 11](https://cloud.githubusercontent.com/assets/17045683/20247351/3c23bf2a-a9ca-11e6-9d04-3210aaaa4854.png)
**Why GBM**:
- competitive performance
- robust
- loss function
- relatively fast

**Usages:**
- quick modeling
- variable selection
- final stage precision modeling

**Competitive performance:**
- competitive with high-end algorithms such as Random Forests
-  reliable performance
- often in winning Kaggle solutions

**Robust:**
- explicitly handle NAs
- normalizing is unnecessary
- handle more factor levels than randomForest (1024 vs 32)
- no limit to the number of independent variables

**Loss functions:**
- Gaussian: squared loss
- Laplace: absolute loss
- Bernouilli: logistic for 0/1
- Huberized: hinge for 0/1
- Adaboost: exponential loss for 0/1
- Multinomial: more than one class (probability matrix)
- Quantile: flexible alpha (optimize for 2stDev threshold)
- Poisson: Poisson distribution for counts
- CoxPH: Cox proportional hazard, for right-censored
- Tdist: t-distribution loss
- Pairwise: ranking (search result scoring)

**Drawbacks:**
- several hyperparameters to tune
- still has capacity to overfit (despite internal cross validation)
- can have trouble with highly noisy data
- black box (GBM package provides tools to analyze results tho)
